File: 6.DP/BOJ/(14501번)퇴사-삼성전자 SW 역량 테스트 기출/[jeunghun] solution.py
n = int(input())
dp = [0] * (n + 1)
t, p = [], []
max_value = 0

for _ in range(n):
    ti, pi = map(int, input().split())

    t.append(ti)
    p.append(pi)

for i in range(n - 1, -1, -1):
    time = i + t[i]

    if time <= n:
        dp[i] = max(p[i] + dp[time], max_value)
        max_value = dp[i]
    else:
        dp[i] = max_value

print(max_value)

File: 6.DP/BOJ/(14501번)퇴사-삼성전자 SW 역량 테스트 기출/[jeunghun] solution.swift
let n = Int(readLine()!)!
var dp = Array(repeating: 0, count: n + 1)
var t = Array<Int>(), p = Array<Int>()
var maxValue = 0

for _ in 0..<n {
    let input = readLine()!.split(separator: " ").map { Int($0)! }
    t.append(input[0])
    p.append(input[1])
}

for i in 0..<n {
    let time = t[n - i - 1] + (n - i - 1)
    
    if time <= n {
        dp[n - i - 1] = max(p[n - i - 1] + dp[time], maxValue)
        maxValue = dp[n - i - 1]
    } else {
        dp[n - i - 1] = maxValue
    }
}

print(maxValue)
